File: medium/0438_find_all_anagrams_in_a_string/solution.py
from typing import List


class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(s) < len(p):
            return []

        k = len(p)

        sCharCounter = [0] * 26
        pCharCounter = [0] * 26

        for i in range(k):
            sCharCounter[ord(s[i]) - 97] += 1
            pCharCounter[ord(p[i]) - 97] += 1

        substrings = []
        if sCharCounter == pCharCounter:
            substrings.append(0)

        for i in range(len(s) - k):
            sCharCounter[ord(s[i]) - 97] -= 1
            sCharCounter[ord(s[i + k]) - 97] += 1

            if sCharCounter == pCharCounter:
                substrings.append(i + 1)

        return substrings

File: medium/0438_find_all_anagrams_in_a_string/solution.cpp
#include <string>
#include <vector>

class Solution {
public:
  std::vector<int> findAnagrams(std::string s, std::string p) {
    if (s.size() < p.size())
      return {};

    const size_t k = p.size();

    std::vector<int> sCharCounter(26, 0);
    std::vector<int> pCharCounter(26, 0);

    size_t i{};
    for (; i < k; ++i) {
      ++sCharCounter[s[i] - 'a'];
      ++pCharCounter[p[i] - 'a'];
    }

    std::vector<int> substrings;
    if (sCharCounter == pCharCounter)
      substrings.push_back(0);

    for (; i < s.size(); ++i) {
      --sCharCounter[s[i - k] - 'a'];
      ++sCharCounter[s[i] - 'a'];

      if (sCharCounter == pCharCounter)
        substrings.push_back(i - k + 1);
    }

    return substrings;
  }
};
